monai/apps/generation/maisi/networks/diffusion_model_unet_maisi.py

```diff
@@ -0,0 +1,411 @@
+            nn.SiLU(),
+            zero_module(
+                Convolution(
+                    spatial_dims=spatial_dims,
+                    in_channels=num_channels[0],
+                    out_channels=out_channels,
+                    strides=1,
+                    kernel_size=3,
+                    padding=1,
+                    conv_only=True,
+                )
+            ),
+        )
+
+    def _create_embedding_module(self, input_dim, embed_dim):
+        model = nn.Sequential(nn.Linear(input_dim, embed_dim), nn.SiLU(), nn.Linear(embed_dim, embed_dim))
+        return model
+
+    def _get_time_and_class_embedding(self, x, timesteps, class_labels):
+        t_emb = get_timestep_embedding(timesteps, self.block_out_channels[0])
+
+        # timesteps does not contain any weights and will always return f32 tensors
+        # but time_embedding might actually be running in fp16. so we need to cast here.
+        # there might be better ways to encapsulate this.
+        t_emb = t_emb.to(dtype=x.dtype)
+        emb = self.time_embed(t_emb)
+
+        if self.num_class_embeds is not None:
+            if class_labels is None:
+                raise ValueError("class_labels should be provided when num_class_embeds > 0")
+            class_emb = self.class_embedding(class_labels)
+            class_emb = class_emb.to(dtype=x.dtype)
+            emb = emb + class_emb
+        return emb
+
+    def _get_input_embeddings(self, emb, top_index, bottom_index, spacing):
+        if self.input_top_region_index:
+            _emb = self.top_region_index_layer(top_index)
+            emb = torch.cat((emb, _emb), dim=1)
+        if self.input_bottom_region_index:
+            _emb = self.bottom_region_index_layer(bottom_index)
+            emb = torch.cat((emb, _emb), dim=1)
+        if self.input_spacing:
+            _emb = self.spacing_layer(spacing)
+            emb = torch.cat((emb, _emb), dim=1)
+        return emb
+
+    def _apply_down_blocks(self, h, emb, context, down_block_additional_residuals):
+        if context is not None and self.with_conditioning is False:
+            raise ValueError("model should have with_conditioning = True if context is provided")
+        down_block_res_samples: list[torch.Tensor] = [h]
+        for downsample_block in self.down_blocks:
+            h, res_samples = downsample_block(hidden_states=h, temb=emb, context=context)
+            for residual in res_samples:
+                down_block_res_samples.append(residual)
+
+        # Additional residual conections for Controlnets
+        if down_block_additional_residuals is not None:
+            new_down_block_res_samples: list[torch.Tensor] = []
+            for down_block_res_sample, down_block_additional_residual in zip(
+                down_block_res_samples, down_block_additional_residuals
+            ):
+                down_block_res_sample = down_block_res_sample + down_block_additional_residual
```

Comment:
You might be able to same some space and time by doing inline operations on tensors you know are fresh and won't be used elsewhere:
```suggestion
                down_block_res_sample.add_(down_block_additional_residual)
```
Here and elsewhere this could save copying a tensor.

Reply:
Hi @ericspod, I have updated this part. Please proceed. Thanks.